codelabs/custom_generator/custom_generator.md

```diff
@@ -0,0 +1,379 @@
+
+```js
+Blockly.defineBlocksWithJsonArray([{
+  "type": "object",
+  "message0": "{ %1 %2 }",
+  "args0": [
+    {
+      "type": "input_dummy"
+    },
+    {
+      "type": "input_statement",
+      "name": "MEMBERS"
+    }
+  ],
+  "output": null,
+  "colour": 230,
+},
+{
+  "type": "member",
+  "message0": "%1 %2 %3",
+  "args0": [
+    {
+      "type": "field_input",
+      "name": "MEMBER_NAME",
+      "text": ""
+    },
+    {
+      "type": "field_label",
+      "name": "COLON",
+      "text": ":"
+    },
+    {
+      "type": "input_value",
+      "name": "MEMBER_VALUE"
+    }
+  ],
+  "previousStatement": null,
+  "nextStatement": null,
+  "colour": 230,
+}]);
+```
+
+### Toolbox definition
+
+Next, define your toolbox in XML. For this example we have a flyout-only toolbox with seven blocks in it.
+
+Copy this code into `custom_generator.js`:
+
+```js
+var codelabToolbox = `
+<xml id="toolbox" style="display: none">
+<block type="object"/>
+<block type="member"></block>
+<block type="math_number"><field name="NUM">0</field></block>
+<block type="text"><field name="TEXT"/></block>
+<block type="logic_boolean"><field name="BOOL">TRUE</field></block>
+<block type="logic_null"/>
+<block type="lists_create_with"><mutation items="3"/></block></xml>
+`
+```
+
+Then update the options struct in the playgound to use your new toolbox:
+
+```js
+var defaultOptions = {
+  comments: true,
+  // ...
+  toolbox: codelabToolbox,
+  // ...
+};
+```
+
+## The basics
+
+The first step is to define and call your generator.
+
+### Create your generator
+
+A custom generator is simply an instance of `Blockly.Generator`. Call the constructor, passing in your generator's name, and store the result.
+
+```js
+const myGenerator = new Blockly.Generator('Codelab');
+```
+
+### Generate code
+
+Add a button to the playground with the following callback:
+
+```js
+const generatedCode = myGenerator.workspaceToCode(Blockly.getMainWorkspace());
+console.log(generatedCode);
+```
+
+Note: You must use `workspaceToCode` as the entry point for code generation, passing in the workspace. `workspaceToCode` calls `init` and `finish` to do any additional setup and teardown that your generator needs.
+
+Other functions such as `blockToCode` are public so that you can use them in your per-block generator definitions, but must not be used as the entry point.
+
+### Test it
+
+Put a number block on the workspace, then click your button to call your generator.
+
+Check the console for the output. You should see an error:
+
+```
+generator.js:182 Uncaught Error: Language "Codelab" does not know how to generate  code for block type "math_number".
+    at Blockly.Generator.blockToCode (generator.js:182)
+    at Blockly.Generator.workspaceToCode (generator.js:94)
+    at <anonymous>:1:18
+```
+
+This error occurs because you need to write a generator for each type of block. Read the next section for more details.
```

Comment:
"generator" in that sentence is a little overloaded as we refer to the whole thing as a generator. Perhaps: "you need to write a code generator for each type of block"

Reply:
https://developers.google.com/blockly/guides/create-custom-blocks/generating-code refers to them as code generators too

Reply:
I've been using "generator" to refer to the instance of Blockly.Generator, and "block generator" for a block's code generation function. Let's discuss here and pick preferred wording, and I'll update through the whole codelab.

Some options:
- "language generator" for the instance of Blockly.Generator, and "block generator" for the individual functions
- "generator", "block generator"
- "language generator", "code generator"
- "generator", "code generator"
- "language generator", "block code generator"
- "generator", "block code generator"

Reply:
"language generator", "code generator" is how I think our documentation refers to these two. Don't really have a strong opinion on this, but maybe: 
"language generator", and "block-to-code generator"?